script/unicode/Unicode/UnicodeSkipList.lean

```diff
@@ -0,0 +1,148 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jean-Baptiste Tristan
+-/
+import Init.Data.Char.UnicodeSkipList
+import Unicode.Unicode
+import Unicode.Parse
+import Init.Data.Char.Basic
+
+open Std System IO FilePath FS
+
+namespace Char.UnicodeSkipList
+
+/-
+The following code creates the unicode skip list data structure for
+a given property. To understand it at a high-level, we provide an
+explanation in Init/Data/Char/UnicodeSkipList.lean
+-/
+
+/-
+Break down the sequence of codepoints into a sequence of ranges,
+alternating between ranges that satisfy the property and ranges that
+do not.
+-/
+def explicitRanges (ucd : List UnicodeData) (property : UnicodeData → Bool) : List Range := Id.run do
+  let mut rangeOpt : Option Range := none
+  let mut ranges := []
+  for datapoint in ucd do
+    let code := datapoint.codepoint
+    let prop := property datapoint
+    match rangeOpt, prop with
+    | some r, true =>
+      if r.stop + 1 = code then
+        -- Extend the range
+        rangeOpt := some {r with stop := code}
+      else
+        -- Hidden gap
+        let completedRange : Range := { start := r.start , stop := r.stop + 1 }
+        let newRange : Range := { start := code , stop := code }
+        rangeOpt := some newRange
+        ranges := completedRange :: ranges
+    | some r, false =>
+      -- Close the range
+      -- Cannot use code for range end as their may be a jump in codepoints
+      let completedRange : Range := { start := r.start , stop := r.stop + 1 }
+      rangeOpt := none
+      ranges := completedRange :: ranges
+    | none, true =>
+      -- Open a range
+      let newRange : Range := { start := code , stop := code }
+      rangeOpt := some newRange
+    | none, false => ()
+
+  return ranges
```

Comment:
This looks like it fails if the last datapoint has the property, because then there is still a pending range in `rangeOpt` that hasn't been added to `ranges` yet.

Reply:
I added some comments to clarify what is going on and what the last case does nothing (and by doing nothing, it just continues to the next codepoint while in a range of codepoints that do not satisfy the property).

Reply:
To give an example of what I mean, I ran
```
#eval (explicitRanges #[UnicodeData.mk "1" 1 (.Letter .Lu)] (fun _ => True)).length
```
and it came out to `0`. I think this is incorrect, I would have expected a list of length one to be returned.

Reply:
Sorry, still catching up here, and haven't seen an answer to @TwoFX last query here.

Reply:
We had a discussion offline. This example has neither a valid unicode database nor a property for which we may want to construct a table, so indeed, the code is not meaningful.